### PR TITLE
v10

### PR DESCRIPTION
**Pull request by timgit**

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -90,90 +81,109 @@ class PgBoss extends EventEmitter {
...
+    if (this.stoppingOn || this.stopped) {
       return
     }
 
-    if (this.stopped) {
-      this.emit(events.stopped)
-    }
-
-    let { destroy = false, graceful = true, timeout = 30000 } = options
+    let { destroy = false, graceful = true, timeout = 30000, wait = false } = options
```
Comment (StarpTech) [inline src/index.js:126]: Can we make `wait=true` the default behavior in v10? I think nobody wants to forcefully exit jobs by default.

File: src/db.js [hunk 2]
```diff
@@ -25,10 +26,49 @@ class Db extends EventEmitter {
...
+      //     console.log(`${new Date().toISOString()}: DEBUG VALUES`)
+      //     console.log(values)
+      //   }
+      // }
+
       return await this.pool.query(text, values)
     }
   }
 
+  async lock ({ timeout = 30, key } = {}) {
```
Comment (StarpTech) [inline src/db.js:43]: I'm a friend of giving variables units. Maybe timeoutSeconds?

File: src/db.js [hunk 3]
```diff
@@ -25,10 +26,49 @@ class Db extends EventEmitter {
 
   async executeSql (text, values) {
     if (this.opened) {
+      // if (this.config.debug === true) {
```
Comment (StarpTech) [inline src/db.js:29]: Can we delete it when not used?

File: types.d.ts [hunk 4]
```diff
@@ -236,7 +229,7 @@ declare namespace PgBoss {
     id: string,
     name: string,
     options: WorkOptions,
-    state: 'created' | 'retry' | 'active' | 'completed' | 'expired' | 'cancelled' | 'failed',
+    state: 'created' | 'retry' | 'active' | 'completed' | 'cancelled' | 'failed',
```
Comment (StarpTech) [inline types.d.ts:232]: Could you add "stopping" here? Then we can complete https://github.com/timgit/pg-boss/issues/422 as well.

**Review by StarpTech — CHANGES_REQUESTED** [review]
Great work! Only a few points.

File: releasenotesv10.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
+v10 is the largest semver major release of pg-boss in years. The API changes included below are ordered by significance.
+
+## Database changes
+PostgreSQL 12 is now the minimum supported version. If you upgrade and run `start()`, the database will automatically be upgraded. However, this release requires rebuilding almost all of the job table indexes, which may require a bit of downtime depending on the size of your queues. If this is a concern, you may extract the migration script via `getMigrationPlans()` and run it against a backup to get an estimate on downtime.
+
+If the standard auto-migration isn't desired, consider alternatives, such as running a new v10 schema side by side of a v9 schema until the v9 queues are drained. 
+
+## API changes
+
+* MAJOR: **Job retries are now opt-out instead of opt-in.** The default `retryLimit` is now 2 retries. This will cause an issue for any job handlers that aren't idempotent. Consider setting `retryLimit=0` on these queues if needed.
```
Comment (StarpTech) [inline releasenotesv10.md:10]: I think this is not a good default. Configure retries when necessary seems to be a better option.

**Comment by StarpTech** [comment]
@timgit are you open to feedback? You resolved all my comments without an answer.

**Comment by KristjanTammekivi** [comment]
Hi, is there any progress here? I would like to have the graceful shutdown to be implemented (#421)

**Comment by timgit** [comment]
Apologies for the long turn around time on this release. There's a few problems I'm still looking at.

**Comment by sneljo1** [comment]
Which items are still open for this?

**Comment by janmeier** [comment]
@timgit Can you put a couple of words on the motivation behind this major version? Which issues does this solve?

I'm saying this with the utmost understanding of what it means to release something as open source - This is your project, and I don't expect you to justify every decision. Still, I'm just curious about the reasoning, and I want to hear about any improvements that should cause us to consider upgrading :)

**Comment by timgit** [comment]
The primary motivation is stability and scalability. In my and others' experience, once a single queue exceeded 1-2mil created jobs in a backlog state, postgres was unable to quickly locate the next record for locking via SKIP LOCKED.  This is mentioned as a limitation in the linked article in the readme.  This produced high memory and CPU utilization issues on the server, which would then impact all other databases. The problem with pg-boss v9's design is that once any queue reaches this failure state, all queues, including internal maintenance queues, were affected since all jobs shared a single table. The design goal was to mitigate a problematic queue by isolating its storage into a dedicated table via partitioning. This doesn't necessarily remove this limitation of SKIP LOCKED, but it should in theory be less of a catastrophic failure if and when it occurs in the future.

Over the years, there were several issues opened around queue concurrency, which added several functions and configuration which made the API more complicated and difficult to understand (`sendSingleton()` and others). There were several ideas that I wanted to prototype, which resulted in queue policies, the biggest feature enhancement worth upgrading for. 

After a couple of failed attempts at building a reasonable migration plan into the new table structure, I decided it was best to release v10 without a schema migration.  This wasn't my first choice, but I think it still was the cleanest way to make it predictable what to expect between these versions.

**Pull request merged**